### PR TITLE
Become plugins

### PR DESCRIPTION
**Pull request by bcoca**
##### SUMMARY

Go from hardcoded become 'methods' to plugins so users can add their own or override the built-in as needed

##### ISSUE TYPE

 - Feature Pull Request
##### COMPONENT NAME

become

### CODE REVIEW

**Comment by acozine** [comment]
@bcoca you can fix this crop of `not-in-toctree` errors by adding `docs/docsite/rst/plugins/become.rst` to match the other plugin-type pages, and including a "glob" TOC:
```
.. toctree:: :maxdepth: 1
    :glob:

    become/*
```

**Comment by bcoca** [comment]
@acozine I was planning on writing that file last, but i'll put in placeholder to avoid the errors

**Comment by jctanner** [comment]
bot_status

File: lib/ansible/plugins/become/runas.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+              - section: runas_become_plugin
+                key: flags
+            vars:
+              - name: ansible_become_flags
+              - name: ansible_runas_flags
+            env:
+              - name: ANSIBLE_BECOME_FLAGS
+              - name: ANSIBLE_RUNAS_FLAGS
+    notes:
+        - runas is really implemented in the winrm conneciton plugin and as such can only be used with winrm connections.
```
Comment (jborean93) [inline lib/ansible/plugins/become/runas.py]: It's more implemented in the powershell shell plugin than winrm. We have both psrp and winrm that support runas as they are both Windows specific. When finally adding support for SSH connections to Windows then it would also be supported there.

**Comment by imthenachoman** [comment]
I do not have a lot of experience with Ansible and I do not have a test environment I can write code in, but I am very familiar with one particular configuration of PowerBroker that uses roles instead of commands invoked via `-u`. If someone will be writing a new PowerBroker plugin I can help explain how it works so they can get it to work with PB roles...

File: lib/ansible/plugins/become/dzdo.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                key: executable
+            vars:
+              - name: ansible_become_exe
+              - name: ansible_dzdo_exe
+            env:
+              - name: ANSIBLE_BECOME_EXE
+              - name: ANSIBLE_DZDO_EXE
+        become_flags:
+            description: Options to pass to dzdo
+            default: -H -S -n
```
Comment (jamescassell) [inline lib/ansible/plugins/become/dzdo.py:43]: Where is this default set?

Comment (sivel) [reply lib/ansible/plugins/become/dzdo.py:43]: Right where you commented.  The `ConfigManger` uses the `DOCUMENTATION` from plugins to configure those plugins.

Comment (jamescassell) [reply lib/ansible/plugins/become/dzdo.py:43]: Awesome!

File: lib/ansible/cli/playbook.py [hunk 3]
```diff
@@ -89,27 +90,26 @@ def run(self):
 
         # initial error check, to make sure all specified playbooks are accessible
         # before we start running anything through the playbook executor
+        pb_dirs = []
         for playbook in self.args:
             if not os.path.exists(playbook):
                 raise AnsibleError("the playbook: %s could not be found" % playbook)
             if not (os.path.isfile(playbook) or stat.S_ISFIFO(os.stat(playbook).st_mode)):
                 raise AnsibleError("the playbook: %s does not appear to be a file" % playbook)
+            pb_dirs.append(os.path.dirname(playbook))
```
Comment (sivel) [inline lib/ansible/cli/playbook.py]: This line is the reason that `The module helloworld_linux was not found in configured module paths` is happening.

`os.path.dirname(playbook)` on something like `test_binary_modules.yml` just becomes `''`.

```suggestion
            pb_dirs.append(os.path.dirname(os.path.abspath(playbook)))
```

Otherwise, you can handle using `abspath` in `add_all_plugin_dirs`

File: lib/ansible/plugins/connection/__init__.py [hunk 4]
```diff
@@ -83,30 +83,30 @@ def __init__(self, play_context, new_stdin, shell=None, *args, **kwargs):
...
-        self._shell = shell_loader.get(shell_type)
+        # helper plugins
+        self._shell = shell
+
+        # we always must have shell
         if not self._shell:
-            raise AnsibleError("Invalid shell type specified (%s), or the plugin for that shell type is missing." % shell_type)
+            self._shell = get_shell_plugin(shell_type=getattr(self, '_shell_type', None), executable=self._play_context.executable)
+
+        self._become = None
```
Comment (sivel) [inline lib/ansible/plugins/connection/__init__.py:90]: Since we are using this outside of the connection (e.g. accessing `getattr(self._connection, '_become')`), we should remove the `_` prefix that indicates it is private.

**Comment by sivel** [comment]
I haven't looked into why exactly, but the `TypeError: 'NoneType' object is not iterable` coming from `ShellBase.chown` is caused by `self._play_context.become_user` being `None` when `_fixup_perms2` is called from `ActionBase._execute_module`.

Because of this `_is_become_unprivileged` is returning `True`, when it shouldn't be.

If I am reading this correctly, I think that we need to use `self._connection._become.get_option('become_user')`

However we have places that expected to be setting `self._play_context.become_user`, so those places may need to update the plugin options.

File: lib/ansible/plugins/connection/__init__.py (hunk 4 above)
Comment (bcoca) [reply lib/ansible/plugins/connection/__init__.py:90]: i would like to remedy that by moving most of those into connection/shell and out of action plugins, which should not care this much about these details.

File: lib/ansible/plugins/become/runas.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+              - section: runas_become_plugin
+                key: user
+            vars:
+              - name: ansible_become_user
+              - name: ansible_runas_user
+            env:
+              - name: ANSIBLE_BECOME_USER
+              - name: ANSIBLE_RUNAS_USER
+            required: True
+        become_exe:
```
Comment (jborean93) [inline lib/ansible/plugins/become/runas.py]: Does this need to be set? We don't call an executable at all with runas so this would do nothing.

File: lib/ansible/plugins/become/runas.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                key: executable
+            vars:
+              - name: ansible_become_exe
+              - name: ansible_runas_exe
+            env:
+              - name: ANSIBLE_BECOME_EXE
+              - name: ANSIBLE_RUNAS_EXE
+        become_flags:
+            description: Options to pass to runas
+            default: -H -S -n
```
Comment (jborean93) [inline lib/ansible/plugins/become/runas.py]: By default we don't set any flags.

File: lib/ansible/plugins/become/runas.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                key: flags
+            vars:
+              - name: ansible_become_flags
+              - name: ansible_runas_flags
+            env:
+              - name: ANSIBLE_BECOME_FLAGS
+              - name: ANSIBLE_RUNAS_FLAGS
+        become_pass:
+            description: password
+            required: False
```
Comment (jborean93) [inline lib/ansible/plugins/become/runas.py]: I assume `required: False` is the default, would it need to be explicitly set?

File: lib/ansible/plugins/become/runas.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    notes:
+        - runas is really implemented in the powershell module handler and as such can only be used with winrm connections.
```
Comment (jborean93) [inline lib/ansible/plugins/become/runas.py:56]: Probably change `winrm` to `Windows` as `psrp`, and soon `ssh`, will support runas against Windows

File: lib/ansible/config/base.yml [hunk 9]
```diff
@@ -1059,50 +1068,6 @@ DEFAULT_SU:
   - {key: su, section: defaults}
   type: boolean
   yaml: {key: defaults.su}
-DEFAULT_SUDO:
-  default: False
-  deprecated:
-    why: In favor of Ansible Become, which is a generic framework
-    version: "2.9"
```
Comment (jborean93) [inline lib/ansible/config/base.yml:1066]: It says this will be removed in 2.9 not 2.8 so sounds like this is being removed a version too early.

File: lib/ansible/plugins/become/runas.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+                key: become_exe
+              - section: runas_become_plugin
+                key: executable
+            vars:
+              - name: ansible_become_exe
+              - name: ansible_runas_exe
+            env:
+              - name: ANSIBLE_BECOME_EXE
+              - name: ANSIBLE_RUNAS_EXE
+        become_flags:
```
Comment (jborean93) [inline lib/ansible/plugins/become/runas.py:29]: While not necessary right now it would be nice if become_flags could take in a dict, or convert key=value, to a dict. Right now become_flags in runas is based on specific key=values and having the controller do the validation vs the remote execution would be nice.

File: lib/ansible/config/base.yml (hunk 9 above)
Comment (jborean93) [reply lib/ansible/config/base.yml:1066]: Ahh I see they have been added to the sudo plugin, ignore me here.

File: lib/ansible/plugins/become/runas.py (hunk 5 above)
Comment (bcoca) [reply lib/ansible/plugins/become/runas.py]: it doesn't, i kept mostly for 'parity' with other plugins

File: lib/ansible/plugins/become/runas.py (hunk 6 above)
Comment (bcoca) [reply lib/ansible/plugins/become/runas.py]: fxied

File: lib/ansible/config/base.yml [hunk 11]
```diff
@@ -329,12 +329,20 @@ ACTION_WARNINGS:
   - {key: action_warnings, section: defaults}
   type: boolean
   version_added: "2.5"
+BECOME_PLUGIN_PATH:
+  name: Become plugins path
+  default: ~/.ansible/plugins/become:/usr/share/ansible/become/action
```
Comment (alikins) [inline lib/ansible/config/base.yml]: `/usr/share/ansible/become/action` ?  should this be /usr/share/ansible/plugins/become ?

File: lib/ansible/config/base.yml [hunk 12]
```diff
@@ -329,12 +329,20 @@ ACTION_WARNINGS:
...
+  description: Colon separated paths in which Ansible will search for Action Plugins.
```
Comment (alikins) [inline lib/ansible/config/base.yml]: s/Action/Become

File: lib/ansible/config/base.yml (hunk 11 above)
Comment (bcoca) [reply lib/ansible/config/base.yml]: good catch, bad edit after copy+paste

File: lib/ansible/config/base.yml (hunk 12 above)
Comment (bcoca) [reply lib/ansible/config/base.yml]: he, same copy paste issue

File: lib/ansible/config/base.yml [hunk 13]
```diff
@@ -329,12 +329,20 @@ ACTION_WARNINGS:
   - {key: action_warnings, section: defaults}
   type: boolean
   version_added: "2.5"
+BECOME_PLUGIN_PATH:
```
Comment (alikins) [inline lib/ansible/config/base.yml:332]: Most of the plugin path config items are of form 'DEFAULT_CACHE_PLUGIN_PATH' etc. 

If there isn't a particular reason for BECOME_PLUGIN_PATH instead of DEFAULT_BECOME_PLUGIN_PATH, may be better to keep it consistent with naming of other plugin path config items.

Comment (bcoca) [reply lib/ansible/config/base.yml:332]: i'm actually  moving away from DEFAULT since that is a misleading tag

File: docs/docsite/Makefile [hunk 14]
```diff
@@ -26,7 +26,7 @@ ifdef PLUGINS
 	PLUGIN_ARGS = -l $(PLUGINS)
 endif
 
-DOC_PLUGINS ?= cache callback connection inventory lookup shell strategy vars
+DOC_PLUGINS ?= become cache callback cliconf connection httpapi inventory lookup shell strategy vars
```
Comment (abadger) [inline docs/docsite/Makefile]: Shouldn't cliconf and httpapi be added in a different PR?  It looks like the landing pages for those two don't currently exist.

File: lib/ansible/cli/__init__.py [hunk 15]
```diff
@@ -339,29 +340,7 @@ def ask_passwords(self):
...
-            self.options.become = True
-            self.options.become_method = 'su'
-            _dep('su')
-
-        # other deprecations:
-        if self.options.ask_sudo_pass or self.options.sudo_user:
-            _dep('sudo')
-        if self.options.ask_su_pass or self.options.su_user:
-            _dep('su')
+        pass
```
Comment (abadger) [inline lib/ansible/cli/__init__.py]: This isn't part of our stable API so we should remove the calls to it and remove the method itself.

File: lib/ansible/executor/task_executor.py [hunk 16]
```diff
@@ -832,12 +831,18 @@ def _get_connection(self, variables, templar):
                     if isinstance(i, string_types) and i.startswith("ansible_") and i.endswith("_interpreter"):
                         variables[i] = delegated_vars[i]
 
-        conn_type = self._play_context.connection
+        # load become plugin if needed
```
Comment (alikins) [inline lib/ansible/executor/task_executor.py:870]: Would it be worthwhile to add a default of no-op become plugin to always have a plugin to avoid the if/elses?
(Could possibly make things like plugins.connection.ConnectionBase.load_become simpler or possibly unneeded)

File: lib/ansible/cli/__init__.py [hunk 17]
```diff
@@ -496,32 +466,18 @@ def base_parser(usage="", output_opts=False, runas_opts=False, meta_opts=False,
...
-
             # consolidated privilege escalation (become)
             runas_group.add_option("-b", "--become", default=C.DEFAULT_BECOME, action="store_true", dest='become',
                                    help="run operations with become (does not imply password prompting)")
-            runas_group.add_option('--become-method', dest='become_method', default=C.DEFAULT_BECOME_METHOD, type='choice', choices=C.BECOME_METHODS,
-                                   help="privilege escalation method to use (default=%s), valid choices: [ %s ]" %
-                                   (C.DEFAULT_BECOME_METHOD, ' | '.join(C.BECOME_METHODS)))
+            runas_group.add_option('--become-method', dest='become_method', default=C.DEFAULT_BECOME_METHOD,
+                                   help="privilege escalation method to use (default=%s), use `ansible-doc -t become -l` to list valid choices."
+                                   % (C.DEFAULT_BECOME_METHOD))
```
Comment (abadger) [inline lib/ansible/cli/__init__.py:472]: We just changed something else to list out the valid methods.  We may want to do the same here.

@gundalow

File: lib/ansible/playbook/block.py [hunk 18]
```diff
@@ -370,7 +370,7 @@ def _get_parent_attribute(self, attr, extend=False, prepend=False):
     def filter_tagged_tasks(self, play_context, all_vars):
```
Comment (abadger) [inline lib/ansible/playbook/block.py]: remove play_context from the method's signature since it's no longer used.

File: lib/ansible/playbook/__init__.py [hunk 19]
```diff
@@ -40,17 +39,18 @@
 
 class Playbook:
 
-    def __init__(self, loader):
+    def __init__(self, loader, options=None):
```
Comment (alikins) [inline lib/ansible/playbook/__init__.py:39]: Passing a command line options instance (a optparse.Values ) deep down into Playbook and Play objects seems wrong to me. It couples ansible.playbook.* stuff to a particular cli arg parser (things like ansible.playbook.FieldAttributeBase._set_option have to deal with the implementation details of optparse)

File: lib/ansible/playbook/play_context.py [hunk 20]
```diff
@@ -446,130 +380,30 @@ def set_task_and_variable_override(self, task, variables, templar):
 
     def make_become_cmd(self, cmd, executable=None):
         """ helper function to create privilege escalation commands """
+        # DEPRECATED
 
-        prompt = None
-        success_key = None
-        self.prompt = None
+        if cmd and self.become:
```
Comment (alikins) [inline lib/ansible/playbook/play_context.py:382]: I'd flip this and return early and deindent the rest of the method

(completely untested example)

``` python
   def make_become_cmd(self, cmd, executable=None):
        if not cmd or not self.become:
            return cmd

        # load/call become plugins here
        plugin = become_loader.get(self.become_method)
        if not plugin:
            raise AnsibleError("Privilege escalation method not found: %s" % self.become_method)

        options = {'become_exe': self.become_exe or getattr(self, '%s_exe' % self.become_method, self.become_method) or self.become_method,
                   'become_flags': self.become_flags or getattr(self, '%s_flags' % self.become_method, '') or '',
                   'become_user': self.become_user,
                   'become_pass': self.become_pass}
        plugin.set_options(direct=options)

        if not executable:
            executable = self.executable

        shell = get_shell_plugin(executable=executable)
        cmd = plugin.build_become_command(cmd, shell)

        # for backwards compat:
        self.prompt = plugin.prompt

        return cmd
```

File: lib/ansible/cli/__init__.py (hunk 15 above)
Comment (bcoca) [reply lib/ansible/cli/__init__.py]: i'm fine with removing it, i was just trying to be 'nice'

File: lib/ansible/executor/task_executor.py (hunk 16 above)
Comment (bcoca) [reply lib/ansible/executor/task_executor.py:870]: I had been thinking about that, a noop plugin is easy enough, the current 'runas' is 99% that.

File: lib/ansible/playbook/block.py (hunk 18 above)
Comment (bcoca) [reply lib/ansible/playbook/block.py]: i kept for backwards compatibility, again its not needed and it is not really 'public api' but wanted to minimize breakage for others.

File: lib/ansible/playbook/__init__.py (hunk 19 above)
Comment (bcoca) [reply lib/ansible/playbook/__init__.py:39]: possibly, this way i saw as 'least intrusive' since now we do it by 'side loading' into play_context, this keeps the correct precedence (cli sets default for keywords).

File: lib/ansible/playbook/play_context.py (hunk 20 above)
Comment (bcoca) [reply lib/ansible/playbook/play_context.py:382]: i prefer to keep 1 return, it seems to lose a lot of people when you have more than one exit from the function

File: lib/ansible/cli/__init__.py (hunk 17 above)
Comment (bcoca) [reply lib/ansible/cli/__init__.py:472]: the other change (list of documentable plugin types) still relied on constants, which won't be the case with plugins

File: docs/docsite/Makefile (hunk 14 above)
Comment (bcoca) [reply docs/docsite/Makefile]: removed from list but added comment

File: lib/ansible/playbook/__init__.py (hunk 19 above)
Comment (bcoca) [reply lib/ansible/playbook/__init__.py:39]: we can change it slightly and make 'options' a dict which is populated from the cli object, but I didn't think we needed that at this point

Comment (abadger) [reply lib/ansible/playbook/__init__.py:39]: @alikins +1

File: lib/ansible/playbook/play_context.py (hunk 20 above)
Comment (abadger) [reply lib/ansible/playbook/play_context.py:382]: @alikins +1  That is better coding style in this case.   If we have a short bit of logic that removes certain specific cases from the rest of the code, it makes sense to return early.  That way we don't have to wonder throughout the rest of the function what happens to the other cases.  https://youtu.be/rrBJVMyD-Gs?t=755

File: lib/ansible/playbook/base.py [hunk 21]
```diff
@@ -231,12 +233,37 @@ def load_data(self, ds, variable_manager=None, loader=None):
                 else:
                     self._attributes[target_name] = ds[name]
 
+        # set defaults from CLI
+        if options:
+            self._set_options(options)
+
```
Comment (abadger) [inline lib/ansible/playbook/base.py:236]: This looks very very fragile.  In addition to what alikins said earlier, we're taking an arbitrary data structure and overwriting attributes in the dictionary if they exist.  If they don't exist, then we're discarding them.  This removes the programming language's builtin abilities to tell when bad data is being passed to it.

This also doesn't seem like it is an inherent part of a become plugin so even if we want to try it, it should be part of a different PR.

File: lib/ansible/playbook/block.py [hunk 22]
```diff
@@ -367,10 +367,10 @@ def _get_parent_attribute(self, attr, extend=False, prepend=False):
 
         return value
 
-    def filter_tagged_tasks(self, play_context, all_vars):
+    def filter_tagged_tasks(self, all_vars):
         '''
         Creates a new block, with task lists filtered based on the tags contained
-        within the play_context object.
+        within the play object. play_context is not in use anymore
```
Comment (abadger) [inline lib/ansible/playbook/block.py]: remove play_context from here too.

File: lib/ansible/playbook/base.py (hunk 21 above)
Comment (bcoca) [reply lib/ansible/playbook/base.py:236]: its would be a prerequisite, since we need a way to get options into the task object from which become consumes.

Comment (abadger) [reply lib/ansible/playbook/base.py:236]: We should be sending specific options in or (better design) the object should be requesting the defaults that it needs.

Comment (bcoca) [reply lib/ansible/playbook/base.py:236]: i agree, but for this change i just copied what we do in play_context, I would leave redoing the API for a subsequent PR ... probably after we remove play_context from the picture.

**Comment by bcoca** [comment]
bot_status

**Comment by bcoca** [comment]
bot_skip

File: lib/ansible/playbook/base.py (hunk 21 above)
Comment (abadger) [reply lib/ansible/playbook/base.py:236]: It probably should go into a PR before this one.

The easiest thing to do is probably to save the CLI options to a program global once it is parsed.  We currently use ansible.constants in that manner so ansible.constants.cli_options might be a suitable name.

**Review by abadger — CHANGES_REQUESTED** [review]

File: lib/ansible/config/base.yml [hunk 23]
```diff
@@ -329,12 +329,20 @@ ACTION_WARNINGS:
...
+  ini:
+  - {key: become_plugins, section: defaults}
+  type: pathspec
 COMMAND_WARNINGS:
   name: Command module warnings
   default: True
   description:
     - By default Ansible will issue a warning when the shell or command module is used and the command appears to be similar to an existing Ansible module.
-    - These warnings can be silenced by adjusting this setting to False. You can also control this at the task level with the module option ``warn``.
+    - These warnings Can be silenced by adjusting this setting to False. You can also control this at the task level with the module option ``warn``.
```
Comment (jimi-c) [inline lib/ansible/config/base.yml]: Typo here.

Comment (bcoca) [reply lib/ansible/config/base.yml]: you spoil my hidden emphasis! .. fixed

File: lib/ansible/playbook/__init__.py (hunk 19 above)
Comment (jimi-c) [reply lib/ansible/playbook/__init__.py:39]: We've talked about converting options to a dict for 3+ years... it would be nice to finally do it.

Comment (bcoca) [reply lib/ansible/playbook/__init__.py:39]: I don't disagree, but i would not make it a requirement here

File: lib/ansible/playbook/play_context.py [hunk 24]
```diff
@@ -443,130 +377,30 @@ def set_task_and_variable_override(self, task, variables, templar):
...
 
-        if self.become:
+            # load/call become plugins here
+            plugin = become_loader.get(self.become_method)
```
Comment (jimi-c) [inline lib/ansible/playbook/play_context.py:385]: I don't understand why we'd keep any of this here anymore, as it could all be done later when the become_loader is used in the connection plugin. With this, we're pretty much just double-loading the plugin (with caching maybe we don't care) but it's just very disjointed, and trying to find where this is done when all the other become stuff has moved out of PlayContext seems confusing.

Comment (bcoca) [reply lib/ansible/playbook/play_context.py:385]: its a backwards compatibility issue, i kept in case 3rd party code still calls this, none of our current code in this PR does so it won't double load for us.

Comment (bcoca) [reply lib/ansible/playbook/play_context.py:385]: it was also a good way to run existing unit tests

File: lib/ansible/constants.py [hunk 25]
```diff
@@ -68,32 +69,14 @@ def set_constant(name, value, export=vars()):
     export[name] = value
 
 
+def BECOME_METHODS():
```
Comment (sivel) [inline lib/ansible/constants.py]: I don't think switching this to a callable, will achieve the result that is expected.  People won't be calling it, so they just wouldn't get anything they expect.

Comment (bcoca) [reply lib/ansible/constants.py]: no, the constant needs to populate from the callable, this is mostly a failed experiment

Comment (bcoca) [reply lib/ansible/constants.py]: the only way to do this with the desired behaviour would be to inject a class as the constants module and have a getter for the specific value ... not sure that is somewhere i want to go.

**Comment by mattclay** [comment]
CI failures/timeouts in sanity, unit and integration tests:

https://app.shippable.com/github/ansible/ansible/runs/99005

**Comment by abadger** [comment]
Being fixed here: https://github.com/ansible/ansible/pull/50421
